Makefile

```diff
@@ -114,11 +114,38 @@ stub-universe-url: docker-dist
 		rm -f $(ROOT_DIR)/stub-universe-url.tmp
 	fi
 
-DCOS_SPARK_TEST_JAR_PATH ?= $(ROOT_DIR)/dcos-spark-scala-tests-assembly-0.2-SNAPSHOT.jar
-$(DCOS_SPARK_TEST_JAR_PATH):
-	cd tests/jobs/scala
-	sbt assembly
-	cp -v $(ROOT_DIR)/tests/jobs/scala/target/scala-2.11/dcos-spark-scala-tests-assembly-0.2-SNAPSHOT.jar $@
+# Special directive to assist in speeding up CI builds
+# Generates a master checksum against the source files in tests/jobs/scala/src/*
+# This is later compared to avoid building identical jars from previous CI runs
+generate-checksum:
+	find tests/jobs/scala/src/* -type f -exec md5sum '{}' + > checklist.chk
+	cut -d ' ' -f1 checklist.chk | sort | md5sum | cut -d ' ' -f1 > $(ROOT_DIR)/master_sum.chk
+
+MD5SUM ?= $(shell cat $(ROOT_DIR)/master_sum.chk)
+DCOS_SPARK_TEST_JAR_VERSION ?= 0.2
+DCOS_SPARK_TEST_JAR_BASE_PATH ?= dcos-spark-scala-tests-assembly-$(DCOS_SPARK_TEST_JAR_VERSION)
+DCOS_SPARK_TEST_JAR_MD5_PATH ?= $(DCOS_SPARK_TEST_JAR_BASE_PATH)-$(MD5SUM).jar
+DCOS_SPARK_TEST_JAR_SNAPSHOT_PATH ?= $(DCOS_SPARK_TEST_JAR_BASE_PATH)-SNAPSHOT.jar
+DCOS_SPARK_TEST_JAR_BASE_URL ?= https://infinity-artifacts.s3.amazonaws.com/autodelete7d/spark/jars
+DCOS_SPARK_TEST_JAR_URL ?= $(DCOS_SPARK_TEST_JAR_BASE_URL)/$(DCOS_SPARK_TEST_JAR_BASE_PATH)-$(MD5SUM).jar
```

Comment:
can we simplify that? it's 6 new variables and I'm not sure that level of flexibility is needed for a jar reference. What if we specify assembly jar name in `build.sbt` without any versions (e.g. `assemblyJarName in assembly := "spark-tests.jar"`) and then add a checksum suffix to it (or no suffix if it's built locally)?

Reply:
Yeah I thought it was becoming a bit overkill. We definitely can simplify but then there's a lot of long links w/ some duplication. The code would look something like this
```
if [ ! -f dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar ]; then
  if [ wget --spider https://infinity-artifacts.s3.amazonaws.com/autodelete7d/spark/jars/dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar 2>/dev/null ]; then
    wget https://infinity-artifacts.s3.amazonaws.com/autodelete7d/spark/jars/dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar
  else
    cd tests/jobs/scala
    sbt assembly
    cp -v target/scala-2.11/dcos-spark-scala-tests-assembly-0.2-SNAPSHOT.jar dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar
    aws s3 cp --acl public-read  \
      s3://infinity-artifacts/autodelete7d/spark/jars/dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar
  fi
fi
cp -v dcos-spark-scala-tests-assembly-0.2-$(MD5SUM).jar $@
```
which doesn't look too bad. I'll fix.